src/protocols/output_power_management.rs

```diff
@@ -0,0 +1,238 @@
+use std::collections::HashMap;
+
+use smithay::{
+    output::Output,
+    reexports::{
+        wayland_protocols_wlr::output_power_management::v1::server::{
+            zwlr_output_power_manager_v1, zwlr_output_power_v1,
+        },
+        wayland_server::{
+            Client, DataInit, Dispatch, DisplayHandle, GlobalDispatch, New, Resource,
+        },
+    },
+};
+use zwlr_output_power_manager_v1::ZwlrOutputPowerManagerV1;
+use zwlr_output_power_v1::ZwlrOutputPowerV1;
+
+const VERSION: u32 = 1;
+
+pub enum Mode {
+    Off,
+    On,
+}
+
+impl From<bool> for Mode {
+    fn from(value: bool) -> Self {
+        match value {
+            false => Self::Off,
+            true => Self::On,
+        }
+    }
+}
+
+impl From<zwlr_output_power_v1::Mode> for Mode {
+    fn from(value: zwlr_output_power_v1::Mode) -> Self {
+        match value {
+            zwlr_output_power_v1::Mode::Off => Self::Off,
+            zwlr_output_power_v1::Mode::On => Self::On,
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl From<Mode> for zwlr_output_power_v1::Mode {
+    fn from(value: Mode) -> Self {
+        match value {
+            Mode::Off => Self::Off,
+            Mode::On => Self::On,
+        }
+    }
+}
+
+pub struct OutputPowerManagerGlobalData {
+    filter: Box<dyn for<'c> Fn(&'c Client) -> bool + Send + Sync>,
+}
+
+pub trait OutputPowerManagementHandler {
+    fn output_power_manager_state(&mut self) -> &mut OutputPowerManagementManagerState;
+
+    // todo: set output power per-monitor. Currently, niri is written to power
+    // all monitors at the same time.
+    fn get_output_power_mode(&mut self, output: &Output) -> Mode;
+    fn set_output_power_mode(&mut self, output: &Output, mode: Mode);
+}
+
+pub struct OutputPowerManagementManagerState {
+    // Active controls only. Failed ones are removed.
+    output_powers: HashMap<Output, ZwlrOutputPowerV1>,
```

Comment:
Something felt off about this, and I think I figured it out.

I based this module off of `gamma_control`. After reading the docs for it, I found out it offers _exclusive_ access to each output, so only one client can create a gamma control for an output. I'm pretty sure that's not true for output power; it should allow multiple clients to connect (operating more like the `output_management` module.